test/spinxsk/spinxsk.c

```diff
@@ -1992,6 +2022,182 @@ ProcessPkts(
+        if (RandUlong() % 2) {
+            //
+            // Add some QUIC flavor to the TCP / UDP payload.
+            //
+            // We want to generate the QUIC header in the TCP/UDP payload buffer
+            // and we use the frame buffer as a random QUIC payload.
+            //
+            payloadLength = payloadBufferSize;
+            UINT16 quicPayloadLength = (UINT16)(RandUlong() % (payloadBufferSize - QUIC_MAX_HEADER_LEN + 1));
+            UINT8 typeSpecificBits = (UINT8)RandUlong();
+            CONST UINT8 version = 1;
+            UINT8 destConnIdLength = (UINT8)(RandUlong() % (XDP_QUIC_MAX_CID_LENGTH + 1));
+            UINT8 destConnId[20];
+            FillRandomBuffer(sizeof(destConnId), destConnId);
+            UINT8 srcConnIdLength = (UINT8)(RandUlong() % (XDP_QUIC_MAX_CID_LENGTH + 1));
+            UINT8 srcConnId[20];
+            FillRandomBuffer(sizeof(srcConnId), srcConnId);
+            BOOLEAN useShortHeader = (BOOLEAN)(RandUlong() % 2);
+            ASSERT_FRE(PktBuildQuicPacket(
+                payload, &payloadLength, frame, quicPayloadLength, typeSpecificBits, version,
+                destConnId, destConnIdLength, srcConnId, srcConnIdLength, useShortHeader));
+        } else {
+            //
+            // Use a random payload.
+            //
+            payloadLength = RandUlong() % (payloadBufferSize + 1);
+        }
+
+        //
+        // Build a transport frame
+        //
+        const UINT32 backfill = RandUlong() % (FNMP_FRAME_MAX_BACKFILL + 1);
+        const UINT16 localPort = (UINT16)RandUlong();
+        const UINT16 remotePort = (UINT16)RandUlong();
+
+        UINT32 frameLength = frameBufferSize - backfill;
+        UINT32 frameType = RandUlong() % 3;
+        if (frameType == 0) {
+            //
+            // Build a UDP frame.
+            //
+            ASSERT_FRE(PktBuildUdpFrame(
+                &frame[backfill], &frameLength, payload, (UINT16)payloadLength,
+                &localMac, &remoteMac, af, &localIp, &remoteIp, localPort, remotePort));
+        } else if (frameType == 1) {
+            //
+            // Build a TCP frame.
+            // Up to 40 bytes of TCP options and random flags.
+            //
+            UINT8 tcpOptions[TCP_MAX_OPTION_LEN];
+            FillRandomBuffer(sizeof(tcpOptions), tcpOptions);
+            UINT16 tcpOptionsLength = 4 * (RandUlong() % (10 + 1));
+            const UINT32 thSeq = RandUlong();
+            const UINT32 thAck = RandUlong();
+            const UINT8 thFlags = (UINT8)RandUlong();
+            const UINT16 thWin = (UINT16)RandUlong();
+
+            ASSERT_FRE(PktBuildTcpFrame(
+                &frame[backfill], &frameLength, payload, (UINT16)payloadLength,
+                tcpOptions, tcpOptionsLength, thSeq, thAck, thFlags, thWin,
+                &localMac, &remoteMac, af, &localIp, &remoteIp, localPort, remotePort));
+        } else {
+            //
+            // Send pure chaos.
```

Comment:
It's also P0 to (sometimes) add chaos to the carefully-constructed frames, because that will be the most effective way of producing nearly-valid yet unexpected data.

Reply:
I can add a pass that goes over the final buffer and modifies bytes randomly with a 1/N probability. That should do the trick. Any opinion for a good N? Roughly the first 100 bytes are relevant for the validity of the input before we have the random payload, so 1/20 would make sense to me.

Reply:
1/20 sounds good